Failed to install the gem (build native extension) on Windows

I'm using ruby 2.4, and I already have ```charlock_holmes``` and ```rugged``` installed. Here is the error message:


Installing github-linguist 6.1.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

current directory:
D:/workdir/vendor/gems/github-linguist-6.1.0/ext/linguist
C:/Ruby24-x64/bin/ruby.exe -r ./siteconf20180515-46084-zv96bg.rb extconf.rb
creating Makefile

current directory:
D:/workdir/vendor/gems/github-linguist-6.1.0/ext/linguist
make "DESTDIR=" clean

current directory:
D:/workdir/vendor/gems/github-linguist-6.1.0/ext/linguist
make "DESTDIR="
generating linguist-x64-mingw32.def
compiling lex.linguist_yy.c
tokenizer.l: In function 'linguist_yylex':
tokenizer.l:96:14: warning: implicit declaration of function 'strndup'
[-Wimplicit-function-declaration]
<sgml>[[:alnum:]_]+=\"            { feed_token(strndup(yytext, strlen(yytext) -
1), REGULAR_TOKEN); eat_until_unescaped('"'); return 1; }
              ^
tokenizer.l:6:23: note: in definition of macro 'feed_token'
     yyextra->token = (tok); \
                       ^~~
tokenizer.l:96:14: warning: incompatible implicit declaration of built-in
function 'strndup'
<sgml>[[:alnum:]_]+=\"            { feed_token(strndup(yytext, strlen(yytext) -
1), REGULAR_TOKEN); eat_until_unescaped('"'); return 1; }
              ^
tokenizer.l:6:23: note: in definition of macro 'feed_token'
     yyextra->token = (tok); \
                       ^~~
tokenizer.l:97:14: warning: incompatible implicit declaration of built-in
function 'strndup'
<sgml>[[:alnum:]_]+='             { feed_token(strndup(yytext, strlen(yytext) -
1), REGULAR_TOKEN); eat_until_unescaped('\''); return 1; }
              ^
tokenizer.l:6:23: note: in definition of macro 'feed_token'
     yyextra->token = (tok); \
                       ^~~
compiling linguist.c
linking shared-object linguist/linguist.so
lex.linguist_yy.o:lex.linguist_yy.c:(.text+0xe2d): undefined reference to
`strndup'
lex.linguist_yy.o:lex.linguist_yy.c:(.text+0x1109): undefined reference to
`strndup'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:259: linguist.so] Error 1

make failed, exit code 2

Gem files will remain installed in
D:/workdir/vendor/gems/github-linguist-6.1.0
for inspection.
Results logged to
D:/workdir/vendor/extensions/x64-mingw32/2.4.0/github-linguist-6.1.0/gem_make.out

An error occurred while installing github-linguist (6.1.0), and Bundler cannot
continue.
Make sure that `gem install github-linguist -v '6.1.0'` succeeds before
bundling.

In Gemfile:
  copyright-header was resolved to 1.0.22, which depends on
    github-linguist

--- a/ext/linguist/linguist.h
+++ b/ext/linguist/linguist.h
@@ -9,3 +9,10 @@ struct tokenizer_extra {
   char *token;
   enum tokenizer_type type;
 };
+
+#ifdef __APPLE__
+char *strndup(const char *s1, size_t n);
+#elif defined(_WIN32) || defined(_WIN64)
+char *strndup(const char *s1, size_t n);
+#pragma warning (disable: 4244)
+#endif // _WIN32 || _WIN64

--- a/ext/linguist/strndup.c
+++ b/ext/linguist/strndup.c
@@ -1,0 +1,20 @@
+#if defined(_WIN32) || defined(_WIN64)
+#include <math.h>
+char *strndup(const char *s1, size_t n)
+{
+	char *copy = (char *)malloc(n + 1);
+	memcpy(copy, s1, n);
+	copy[n] = 0;
+	return copy;
+}
+#endif
+
+
+#if defined(__APPLE__)
+char *strndup(const char *s1, size_t n)
+{
+  char *copy = calloc(n + 1, sizeof(char));
+  memcpy(copy, s1, n);
+  return copy;
+}
+#endif

